### DISCUSSION TITLE
How to update the integrated ControlNET ?

### DISCUSSION THREAD

**Discussion by TomiTom1234** [lllyasviel/stable-diffusion-webui-forge#178] [Q&A] [answered]
Hi!

I just wanted to try "ip-Adapter" in ControlNet, but I can't find the "ip-adapter_face_id_plus" Preprocessor because it is not updated to the last version. How can I update CN to the last version?
I tried to intstall CN also but it keeps disabled because there is CN already integrated. Any idea please?

**Comment by 2blackbar** [comment]
https://github.com/lllyasviel/stable-diffusion-webui-forge/discussions/97

**Comment by TomiTom1234** [comment]
> #97

Nope, this is something different. I am not talking about the models, which I already saved in the ControlNET models folder, I am talking here about the preprocessor, which is not in the integrated CN because it is not on the last version.

**Reply by 2blackbar** [reply]
oh well... faceid works but its pretty meh...  
![image](https://github.com/lllyasviel/stable-diffusion-webui-forge/assets/33694664/5ed5a3b0-c16c-48ce-8bfd-c17ec44c1da6)


So... use instantID instead

![image](https://github.com/lllyasviel/stable-diffusion-webui-forge/assets/33694664/b12e72eb-7ba4-450e-9aef-e9f9bfce9df7)

Faceidportrait is so so...
![image](https://github.com/lllyasviel/stable-diffusion-webui-forge/assets/33694664/aac10c72-4aec-4458-96f7-fcfb8c2e6cbf)

![image](https://github.com/lllyasviel/stable-diffusion-webui-forge/assets/33694664/f4bde917-a928-4dee-9512-dcaa0cb7c9c0)

**Reply by TomiTom1234** [reply]
Thanks for the tips.
But did you notice that there are some Preprocessors messing? I assume the CN is not updated, or is there a way to download them and put them in a folder inside Forge?

**Reply by 2blackbar** [reply]
yes but they do work with models , they are just named differently, you have to figure out which ones work with what models, i did not look around where preprocessors are stored , maybe somewhere  FORGE\webui\extensions-builtin   OR  FORGE\webui\models\ControlNetPreprocessor

**Reply by TomiTom1234** [reply]
Yes this is the right folder FORGE\webui\models\ControlNetPreprocessor , but where is that preprocessor to download from? I know the models to download, usually the ControlNET get the Preprocessors automatically when the CN at last version, so if there is that preprocessor to download I would love from anyone to give me a direct link for that.

**Reply by 2blackbar** [reply]
ok, looks like a lot ipadapters does not work , throws ouyt black image, i think its incomplete, and you cant just swap regular controlnet  cause it does not work , at least for me , i think you should creae new issue , maybe it was not tested by dev
      File "F:\FORGE\webui\modules\call_queue.py", line 57, in f
        res = list(func(*args, **kwargs))
    TypeError: 'NoneType' object is not iterable

**Reply by Moussack** [reply]
FaceID works pretty well for me. inside forge's CN the preprecesor is called "**InsightFace+CLIP-H (IPAdapter)**" 

**CONTROLNET:**
![NVIDIA_Share_QEYBEj1JDO](https://github.com/lllyasviel/stable-diffusion-webui-forge/assets/69501089/3980cd3b-a298-488c-91f5-4cdb05499a29)

**INPUT IMAGE :**
![303871568-b12e72eb-7ba4-450e-9aef-e9f9bfce9df7](https://github.com/lllyasviel/stable-diffusion-webui-forge/assets/69501089/58ea7b3d-49fd-41ce-9983-2f87397171ba)

**RESULTS :**
![BL](https://github.com/lllyasviel/stable-diffusion-webui-forge/assets/69501089/30868918-0ffd-4a8f-85be-d3233361a8b6)
![00219-2869513814](https://github.com/lllyasviel/stable-diffusion-webui-forge/assets/69501089/53abe6a2-c682-4757-9277-f93cf37c182b)
![BL3](https://github.com/lllyasviel/stable-diffusion-webui-forge/assets/69501089/2ce7386c-4e06-4277-b9c6-6374f2e96daf)
![00248-1896951166](https://github.com/lllyasviel/stable-diffusion-webui-forge/assets/69501089/bdf4e354-bd36-4faf-821d-ea69c61a6066)
![00243-4148844535](https://github.com/lllyasviel/stable-diffusion-webui-forge/assets/69501089/e3cce11a-1c82-4834-8d7c-049803341748)

**Reply by TomiTom1234** [reply]
Yes you are right, it is the only option as a Preprocessor for IP-Adapter, and as you said it works well, but still there are some messing IP-Adapters which I mentioned, which - maybe - means that the CN is not updated to the last version and there is no way to update it (I think) by ourselves.
I wish someone corrects me if I am wrong.

**Reply by nharwell** [reply]
Issue 179 is open on this problem - https://github.com/lllyasviel/stable-diffusion-webui-forge/issues/179
Scroll to the screenshot & comments at the bottom.

**Reply by hoodady** [reply]
I donno but im getting really poor results:
![image](https://github.com/lllyasviel/stable-diffusion-webui-forge/assets/76712106/52639e3a-e758-41d9-8718-23bdca9cfd90)

**Reply by a3nima** [reply]
@hoodady  dont forget to use the LORA ;-)

**Comment by CRCODE22** [comment]
Why does the built in control net not allow multiple input images? The external control net being used by automatic 1111 does allow that. When I install that extension in forge it will not allow you to use it because it keeps disabling it automatically even when I check it to be enabled.

**Comment by CRCODE22** [comment]
The main controlnet is much better than the build in one because the build in one does not have multi input feature which the main branch has which you can see below but forge will not allow you to use it but automatic 1111 does.

![multiinputs](https://github.com/lllyasviel/stable-diffusion-webui-forge/assets/88407346/5b3747e8-eacc-42e1-96c2-6286795142b8)

**Reply by a3nima** [reply]
in forge you can use a mask and put 2 characters on one image. not possible with A1111. BUT still no multi image input in main branch of forge :(

**Reply by nekrophiliaki** [reply]
Do you have more information on how to do this? I have used Composable Lora with poor results.

**Comment by fonik000** [comment]
I have a temporary alternative: 
1. Download the original controlnet.
2. To be on the safe side, make a copy of the folder: sd_forge_controlnet
3. Copy the files of the original controlnet into the folder: sd_forge_controlnet and overwrite all files. 
4. Now you have the latest version of controlnet. You will now see face-id as the preprocessor.

**Reply by the0ckid82** [reply]
did this work for you ?

**Reply by TomiTom1234** [reply]
Somehow the recommendation of [fonik000](https://github.com/fonik000) worked to show the exact same options and preprocessors that the original CN has, but there were some errors here and there, so I decided to go back to the integrated CN, and to be honest after testing I see that the pre-installed preprocessor in this integrated CN "InsightFace+CLIP-H (IPAdapter)" does pretty good - if not better - job than the "ip-adapter_face_id_plus" preprocessor.

**Reply by a3nima** [reply]
its not preinstalled, it has to be downloaded ;) but you're right the preprocessor you mentioned works well :)

**Reply by escmaster** [reply]
Thank you so much, I'm already try another version of SD but with your instruction, I can go back using SD-Forge!!! Works like a charm for me!! Even fix a lot of errors that I'm getting from controlnet maybe because was outdated.

**Comment by lllyasviel** [comment, marked as answer]
Hi please note that the previous naming of ipadapters preprocessors in `sd-webui-controlnet` were wrong.
The names in Forge are correct and consistent to official naming from ipadapter official :

| SD v. | IPadapter | Img encoder |
|---|---|---|
| v1.5 | [ip-adapter_sd15](https://huggingface.co/h94/IP-Adapter/resolve/main/models/ip-adapter_sd15.safetensors) | ViT-H | 
| v1.5 | [ip-adapter_sd15_light](https://huggingface.co/h94/IP-Adapter/resolve/main/models/ip-adapter_sd15_light.safetensors) | ViT-H | 
| v1.5 | [ip-adapter-plus_sd15](https://huggingface.co/h94/IP-Adapter/resolve/main/models/ip-adapter-plus_sd15.safetensors) | ViT-H | 
| v1.5 | [ip-adapter-plus-face_sd15](https://huggingface.co/h94/IP-Adapter/resolve/main/models/ip-adapter-plus-face_sd15.safetensors) | ViT-H | 
| v1.5 | [ip-adapter-full-face_sd15](https://huggingface.co/h94/IP-Adapter/resolve/main/models/ip-adapter-full-face_sd15.safetensors) | ViT-H | 
| v1.5 | [ip-adapter_sd15_vit-G](https://huggingface.co/h94/IP-Adapter/resolve/main/models/ip-adapter_sd15_vit-G.safetensors) | ViT-bigG | 
| SDXL | [ip-adapter_sdxl](https://huggingface.co/h94/IP-Adapter/resolve/main/sdxl_models/ip-adapter_sdxl.safetensors) | ViT-bigG | 
| SDXL | [ip-adapter_sdxl_vit-h](https://huggingface.co/h94/IP-Adapter/resolve/main/sdxl_models/ip-adapter_sdxl_vit-h.safetensors) | ViT-H | 
| SDXL | [ip-adapter-plus_sdxl_vit-h](https://huggingface.co/h94/IP-Adapter/resolve/main/sdxl_models/ip-adapter-plus_sdxl_vit-h.safetensors) | ViT-H | 
| SDXL | [ip-adapter-plus-face_sdxl_vit-h](https://huggingface.co/h94/IP-Adapter/resolve/main/sdxl_models/ip-adapter-plus-face_sdxl_vit-h.safetensors) | ViT-H | 
| v1.5 | [FaceID](https://huggingface.co/h94/IP-Adapter-FaceID/resolve/main/ip-adapter-faceid_sd15.bin) | InsightFace+CLIP-H | 
| v1.5 | [FaceID Plus](https://huggingface.co/h94/IP-Adapter-FaceID/resolve/main/ip-adapter-faceid-plus_sd15.bin) | InsightFace+CLIP-H | 
| v1.5 | [FaceID Plus v2](https://huggingface.co/h94/IP-Adapter-FaceID/resolve/main/ip-adapter-faceid-plusv2_sd15.bin) | InsightFace+CLIP-H | 
| v1.5 | [FaceID Portrait](https://huggingface.co/h94/IP-Adapter-FaceID/resolve/main/ip-adapter-faceid-portrait_sd15.bin) | InsightFace+CLIP-H |
| SDXL | [FaceID](https://huggingface.co/h94/IP-Adapter-FaceID/resolve/main/ip-adapter-faceid_sdxl.bin) | InsightFace+CLIP-H |
| SDXL | [FaceID Plus v2](https://huggingface.co/h94/IP-Adapter-FaceID/resolve/main/ip-adapter-faceid-plusv2_sdxl.bin) | ViT-H |

**Reply by TigerHH6866** [reply]
how about making forge CN and sd-webui-controlnet same name of preprocessor

**Reply by Moussack** [reply]
That's what I'm thinking,the faceid preprocessor is the InsightFace+CLIP-H one, cuz it gives the great results on the Bruce Lee example above, Thx for the clarification.

**Reply by MartialWrapperr** [reply]
I'm getting " InsightFace must be provided for FaceID models." with FaceID Plus v2, BTW.

**Reply by Asmus-code** [reply]
The FaceID Plus v2 says to use ViT-H, but when you use it, it yells saying you need insightface, like Martial above is stating. Not sure if ViT was listed incorrectly?

**Reply by Moussack** [reply]
@Asmus-code are you using it on SDXL or SD1.5??

**Reply by abline11** [reply]
I get the InsightFace+CLIP-H (IPAdapter)/ip-adapter-faceid-plusv2_sdxl [187cb962] combination to work perfectly.  However there is an annoying 'sub-bug' that throws up the message:

*** Error running process_before_every_sampling: C:\AI\stable-diffusion-webui-forge\webui\extensions-builtin\sd_forge_controlnet\scripts\controlnet.py
    Traceback (most recent call last):
      File "C:\AI\stable-diffusion-webui-forge\webui\modules\scripts.py", line 835, in process_before_every_sampling
        script.process_before_every_sampling(p, *script_args, **kwargs)
      File "C:\AI\stable-diffusion-webui-forge\system\python\lib\site-packages\torch\utils\_contextlib.py", line 115, in decorate_context
        return func(*args, **kwargs)
      File "C:\AI\stable-diffusion-webui-forge\webui\extensions-builtin\sd_forge_controlnet\scripts\controlnet.py", line 555, in process_before_every_sampling
        self.process_unit_before_every_sampling(p, unit, self.current_params[i], *args, **kwargs)
    KeyError: 0

... this is indicating that before every sampling it is checking the contents of all the ControlNets (by default ControlNet 0, 1 and 2.

If you are using just one with InsightFace+CLIP-H (IPAdapter)/ip-adapter-faceid-plusv2_sdxl [187cb962], you will also see that what's in ControlNet 0 has also been copied down into ControlNets 1 and 2. If you 'un' enable ControlNets 1 and 2 and set the control types in those to ALL thus setting the preprocessor and model combination back to none/none for ControlNets 1 and 2, you will see the error message goes away.

So that's how to get rid of the message. Unfortunately, if you go into settings to try and set this as the Default it will not do so and next time you completely restart Forge from scratch the Controlnets 1 and 2 will be back to what's in ControlNet 0 and the error message will reappear.

It's a shame that making ControlNet a built in extension is causing some of these issues that don't appear in Automatic1111, but I really appreciate the hard work that has been put in.

**Reply by Vincentdu-cn** [reply]
If you use IP-Adapter with CLIP-ViT-H (IPAdapter),when I run preprocessor,every time it will download model.safetensors from huggingface,but I have already download the file,and rename CLIP-ViT-H-14.safetensors.
Then I use InsightFace+CLIP-H (IPAdapter),it will also download the same file.

Finally got stuck at this download step.

```shell
Begin to load 1 model
Moving model(s) has taken 0.02 seconds
Model loaded in 16.1s (load weights from disk: 2.1s, forge load real models: 11.5s, forge finalize: 0.1s, load VAE: 0.2s, calculate empty prompt: 1.9s).
2024-03-17 14:57:06,498 - ControlNet - INFO - Preview Resolution = 512
Downloading: "https://huggingface.co/h94/IP-Adapter/resolve/main/models/image_encoder/model.safetensors" to /home/stable-diffusion-webui-forge/models/ControlNetPreprocessor/CLIP-ViT-H-14.safetensors
```

**Reply by ely1113** [reply]
Where can I download InsightFace+CLIP-H? I can't find it anywhere.

**Reply by Vincentdu-cn** [reply]
https://huggingface.co/h94/IP-Adapter/resolve/main/models/image_encoder/model.safetensors?download=true

**Reply by DiegoRRR** [reply]
> SD v.	IPadapter	Img encoder (...)

In which folder do I put these files? I put them in "ControlNetPreprocessor" but they don't appear.

**Reply by ssbugman** [reply]
> Hi please note that the previous naming of ipadapters preprocessors in `sd-webui-controlnet` were wrong. The names in Forge are correct and consistent to official naming from ipadapter official :
> 
> SDXL	[FaceID Plus v2](https://huggingface.co/h94/IP-Adapter-FaceID/resolve/main/ip-adapter-faceid-plusv2_sdxl.bin)	ViT-H

ViT-H requires Insightface howi nstall it?
InsightFace+CLIP-H dont working with SDXL models and returns messed mutated face.

**Comment by gohan2091** [comment]
So in XL, if I want to use FaceID Plus v2, I need to select ViT-H? because when I do that, the face doesn't transfer to the image, it's ignored. I have to use InsightFace+CLIP-H and then it works

**Comment by foychermin** [comment]
is there a fix for openpose and forges controlnet yet?

**Comment by TheCodeSlinger** [comment]
Here I made a multi-input for when using IpAdapter in the built in control net. I'm still not impressed with the results compared to the official control net version when using Automatic1111's, but maybe someone wants to look at the code and has some suggestions. 

https://github.com/TheCodeSlinger/ForgeUI-IPAdapter-MultiInput
